lib/std/priority_dequeue.zig

```diff
@@ -0,0 +1,907 @@
+                }
+            }
+
+            const second_grandchild_index = first_grandchild_index + 1;
+            if (second_grandchild_index >= self.len) {
+                // One grandchild, so we know there is a second child. Compare first grandchild and second child
+                return self.bestItemAtIndices(first_grandchild_index, second_child_index, compare);
+            }
+
+            const best_left_grandchild_index = self.bestItemAtIndices(first_grandchild_index, second_grandchild_index, compare).index;
+            const third_grandchild_index = second_grandchild_index + 1;
+            if (third_grandchild_index >= self.len) {
+                // Two grandchildren, and we know the best. Compare this to second child.
+                return self.bestItemAtIndices(best_left_grandchild_index, second_child_index, compare);
+            } else {
+                // Three grandchildren, compare the min of the first two with the third
+                return self.bestItemAtIndices(best_left_grandchild_index, third_grandchild_index, compare);
+            }
+        }
+
+        /// Return the number of elements remaining in the dequeue
+        pub fn count(self: Self) usize {
+            return self.len;
+        }
+
+        /// Return the number of elements that can be added to the
+        /// dequeue before more memory is allocated.
+        pub fn capacity(self: Self) usize {
+            return self.items.len;
+        }
+
+        /// Dequeue takes ownership of the passed in slice. The slice must have been
+        /// allocated with `allocator`.
+        /// De-initialize with `deinit`.
+        pub fn fromOwnedSlice(allocator: *Allocator, lessThanFn: fn (T, T) bool, items: []T) Self {
+            var queue = Self{
+                .items = items,
+                .len = items.len,
+                .allocator = allocator,
+                .lessThanFn = lessThanFn,
+            };
+
+            if (queue.len <= 1) return queue;
+
+            const half = (queue.len >> 1) - 1;
+            var i: usize = 0;
+            while (i <= half) : (i += 1) {
+                const index = half - i;
+                queue.siftDown(index);
+            }
+            return queue;
+        }
+
+        pub fn ensureCapacity(self: *Self, new_capacity: usize) !void {
+            var better_capacity = self.capacity();
+            if (better_capacity >= new_capacity) return;
+            while (true) {
+                better_capacity += better_capacity / 2 + 8;
+                if (better_capacity >= new_capacity) break;
+            }
+            self.items = try self.allocator.realloc(self.items, better_capacity);
+        }
+
+        pub fn resize(self: *Self, new_len: usize) !void {
+            try self.ensureCapacity(new_len);
+            self.len = new_len;
+        }
+
+        pub fn shrink(self: *Self, new_len: usize) void {
+            // TODO take advantage of the new realloc semantics
```

Comment:
I copied this straight from `std.PriorityQueue`. It's not a great TODO, I am not sure exactly what it's referring to. It would be nice to fix this in both places while I am touching this code, but I will need someone to point me in the right direction.

Reply:
Check out `Allocator.realloc`, the reallocation can fail whenever the backing allocator deems it unnecessary (while `shrink` is guaranteed to get you a smaller piece of memory).
This scheme is already in use in `ArrayList.shrinkAndFree`.

Reply:
Thanks. I have replaced `shrink` with `shrinkRetainingCapacity` and `shrinkAndFree`. In both cases I have restricted the new length to be between the length and the capacity, as shrinking to a smaller size would invalidate the heap property. An alternative would be to remove the largest `n` items instead, thus retaining the property - I think what I have done is probably the safest default behaviour but I can see arguments both ways. This forces the user to decide which elements they want to remove if they want to shrink the queue below its current size.